File: 73_SetMatrixZeroes.java
//O(n+m)
/*
class Solution {
    public void setZeroes(int[][] matrix) {
        int row = matrix.length;
        int col = matrix[0].length;
        
        boolean[] rowFlags = new boolean[row];
        boolean[] colFlags = new boolean[col];
        
        for(int i = 0 ; i < row; i++) {
        	for(int j = 0; j < col; j++) {
        		if(matrix[i][j] == 0) {
        			rowFlags[i] = true;
                    colFlags[j] = true;
        		}
        	}
        }
        
        for (int i = 0; i < row; i++) {
            if (rowFlags[i]) {
                for (int j = 0; j < col; j++) {
                    matrix[i][j] = 0;
                }
            }
        }

        for (int i = 0; i < col; i++) {
            if (colFlags[i]) {
                for (int j = 0; j < row; j++) {
                    matrix[j][i] = 0;
                }
            }
        }
    }
}
*/
public void setZeroes(int[][] matrix) {
        int noZeroRawIndex = -1;
        int rawLength = matrix.length;
        int colLength = matrix[0].length;
        for (int i = 0; i < rawLength; i++) {
            int index = 0;
            for (index = 0; index < colLength; index++) {
                if (matrix[i][index] == 0) {
                    break;
                }
            }
            if (index == colLength) {
                noZeroRawIndex = i;
                break;
            }
        }

        if (noZeroRawIndex == -1) {
            for (int i = 0; i < rawLength; i++) {
                for (int j = 0; j < colLength; j++) {
                    matrix[i][j] = 0;
                }
            }
        } else {
            for (int i = 0; i < rawLength; i++) {
                for (int j = 0; j < colLength; j++) {
                    if (matrix[i][j] == 0) {
                        matrix[noZeroRawIndex][j] = 0;
                    }
                }
            }

            for (int i = 0; i < rawLength; i++) {
                for (int j = 0; j < colLength; j++) {
                    if (i == noZeroRawIndex) {
                        continue;
                    }
                    if (matrix[i][j] == 0) {
                        for (int j2 = 0; j2 < colLength; j2++) {
                            matrix[i][j2] = 0;
                        }
                    }
                }
            }

            for (int i = 0; i < colLength; i++) {
                if (matrix[noZeroRawIndex][i] == 0) {
                    for (int j = 0; j < rawLength; j++) {
                        matrix[j][i] = 0;
                    }
                }
            }
        }
    }
}

File: c++/73_SetMatrixZeroes.cpp
class Solution {
public:
    void setZeroes(vector<vector<int>>& matrix) {
        int m = (int)matrix.size();
        int n = (int)matrix[0].size();
        
        vector<int> row(m), col(n);
        
        for (int i = 0; i < m; i++) {
            for (int j = 0; j < n; j++) {
                if (matrix[i][j] == 0) {
                    row[i] = col[j] = true;
                }
            }
        }
        
        for (int i = 0; i < m; i++) {
            for (int j = 0; j < n; j++) {
                if (row[i] || col[j]) {
                    matrix[i][j] = 0;
                }
            }
        }
    }
};

class Solution {
public:
    void setZeroes(vector<vector<int>>& matrix) {
        int m = (int)matrix.size();
        int n = (int)matrix[0].size();
        
        int flag_col0 = false, flag_row0 = false;
        
        for (int i = 0; i < m; i++) {
            if (matrix[i][0] == 0) {
                flag_col0 = true;
            }
        }
        for (int j = 0; j < n; j++) {
            if (matrix[0][j] == 0) {
                flag_row0 = true;
            }
        }
        
        for (int i = 1; i < m; i++) {
            for (int j = 1; j < n; j++) {
                if (matrix[i][j] == 0) {
                    matrix[i][0] = matrix[0][j] = 0;
                }
            }
        }
        
        for (int i = 1; i < m; i++) {
            for (int j = 1; j < n; j++) {
                if (matrix[i][0] == 0 || matrix[0][j] == 0) {
                    matrix[i][j] = 0;
                }
            }
        }
        
        if (flag_col0) {
            for (int i = 0; i < m; i++) {
                matrix[i][0] = 0;
            }
        }
        if (flag_row0) {
            for (int j = 0; j < n; j++) {
                matrix[0][j] = 0;
            }
        }
    }
};


